src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+ */
+
+SUNAdaptController SUNAdaptController_I(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_I(C, DEFAULT_I_K1);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PI parameters
+ */
+
+int SUNAdaptController_SetParams_I(SUNAdaptController C, sunrealtype k1)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = RCONST(0.0);
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an explicit Gustafsson controller
+ */
+
+SUNAdaptController SUNAdaptController_ExpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ExpGus(C,
+                                             DEFAULT_EXPGUS_K1,
+                                             DEFAULT_EXPGUS_K2);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set explicit Gustafsson parameters
+ */
+
+int SUNAdaptController_SetParams_ExpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1+k2;
+  SODERLIND_K2(C) = -k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create an implicit Gustafsson controller
+ */
+
+SUNAdaptController SUNAdaptController_ImpGus(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_ImpGus(C,
+                                             DEFAULT_IMPGUS_K1,
+                                             DEFAULT_IMPGUS_K2);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set explicit Gustafsson parameters
+ */
+
+int SUNAdaptController_SetParams_ImpGus(SUNAdaptController C,
+                                        sunrealtype k1, sunrealtype k2)
+{
+  SODERLIND_K1(C) = k1+k2;
+  SODERLIND_K2(C) = -k2;
+  SODERLIND_K3(C) = RCONST(0.0);
+  SODERLIND_K4(C) = RCONST(1.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+
+
+/* -----------------------------------------------------------------
+ * implementation of controller operations
+ * ----------------------------------------------------------------- */
+
+SUNAdaptController_Type SUNAdaptController_GetType_Soderlind(SUNAdaptController C)
+{ return SUN_ADAPTCONTROLLER_H; }
+
+int SUNAdaptController_EstimateStep_Soderlind(SUNAdaptController C, sunrealtype h,
+                                              int p, sunrealtype dsm, sunrealtype* hnew)
+{
```

Comment:
```suggestion
{
  if (C == NULL || hnew == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.